File: ARC122/a.cc
#include <iostream>
#include <cmath>
#include <algorithm>
#include <vector>
using namespace std;

long int a[100005];
long long int m = pow(10, 9)+7;


int main(){
  long long int n, sum=0, ans=0, size, hiku, tmp, i, count;
  std::cin >> n;
  for(i =0; i<n; ++i){
    std::cin >> a[i];
    sum = (sum+a[i])%m;
  }


hiku = (sum - a[0])%m;
sum = sum%m;
ans = sum;

ans += (ans *(n-1)-hiku*2)%m;



size = n/2;
i = 2;
while(size >1){
  tmp =0;
  count = 0;
  int lef= n-1-i, rig=n-1;

  while(lef != 0){
    tmp += a[lef]+a[rig];
    lef--;
    rig--;
    count ++;
  }
  std::cout << size << '\n';

  ans = ((ans- tmp*2)%m + (sum*count)%m)%m;

  size = size/2;

  //std::cout << "ans: " << ans << '\n';
}


std::cout << ans << '\n';


  return 0;
}

File: ARC122/a.py
# -*- coding: utf-8 -*-
import numpy as np

m = 10**9+7

n = int(input())

sum =0;


hoge = list(map(int, input().split()))


for i in range(n):
    sum += hoge[i]

hiku = (sum-hoge[0])

ans = sum
ans += ans*(n-1)-hiku*2

size = n/2
i=2

while size>1:
    tmp = 0
    count = 0
    left = n-1-i
    right = n-1
    
    while left != 0:
        tmp += hoge[left]+ hoge[right]
        left-=1
        right-=1
        count+=1
    
    ans += (sum*count-(tmp*2))%m
    size = int(size/2)
    i+=1

print(ans%m)
        